### PR TITLE
Output accuracies per class.

### PR DESCRIPTION
**Pull request by rmanor**

### CODE REVIEW

File: src/caffe/layers/accuracy_layer.cpp [hunk 1]
```diff
@@ -82,6 +90,11 @@ void AccuracyLayer<Dtype>::Forward_cpu(const vector<Blob<Dtype>*>& bottom,
 
   // LOG(INFO) << "Accuracy: " << accuracy;
   top[0]->mutable_cpu_data()[0] = accuracy / count;
+  if (top.size() > 1) {
+    for (int i = 0; i < dim; ++i) {
+      top[1]->mutable_cpu_data()[i] = accuracies[i] / nums[i];
```
Comment (ronghanghu) [inline src/caffe/layers/accuracy_layer.cpp]: What if you get 0/0 here?

Comment (rmanor) [inline src/caffe/layers/accuracy_layer.cpp]: Thanks, fixed.

File: src/caffe/layers/accuracy_layer.cpp [hunk 2]
```diff
@@ -50,11 +54,14 @@ void AccuracyLayer<Dtype>::Forward_cpu(const vector<Blob<Dtype>*>& bottom,
   const int num_labels = bottom[0]->shape(label_axis_);
   vector<Dtype> maxval(top_k_+1);
   vector<int> max_id(top_k_+1);
+  vector<Dtype> accuracies(dim, 0);
+  vector<Dtype> nums(dim, 0);
```
Comment (ronghanghu) [inline src/caffe/layers/accuracy_layer.cpp]: Maybe it is better to keep these two vectors `accuracies` and `nums` persistent as a member variable in AccuracyLayer?

Comment (rmanor) [inline src/caffe/layers/accuracy_layer.cpp]: Why?
They are usually small and it doesn't cost much to create them.
Also, they do not represent a state of the class so I don't think they should be class variables.

Comment (ronghanghu) [inline src/caffe/layers/accuracy_layer.cpp]: My opinion is to keep these two variables as computation buffers and put them as member variable blobs as is done in many other layers (actually you only need to create one such vector and use top[1] itself as another). The advantages are
* you don't need to reallocate them every time.
* you don't need to allocate them if `top.size() == 1`

>They are usually small and it doesn't cost much to create them.

Not necessarily true, sometimes we may want to do 10,000+ category classification in caffe.

**Comment by ronghanghu** [comment]
You also need to address `ignore_label` https://github.com/BVLC/caffe/blob/master/src/caffe/proto/caffe.proto#L413-L414 and skip them in per-class accuracy.

File: src/caffe/layers/accuracy_layer.cpp [hunk 3]
```diff
@@ -50,11 +54,14 @@ void AccuracyLayer<Dtype>::Forward_cpu(const vector<Blob<Dtype>*>& bottom,
...
   int count = 0;
   for (int i = 0; i < outer_num_; ++i) {
     for (int j = 0; j < inner_num_; ++j) {
       const int label_value =
           static_cast<int>(bottom_label[i * inner_num_ + j]);
+      ++nums[label_value];
       if (has_ignore_label_ && label_value == ignore_label_) {
         continue;
       }
```
Comment (ronghanghu) [inline src/caffe/layers/accuracy_layer.cpp]: move `++nums[label_value];` (3 lines above) after this line.

**Comment by bhack** [comment]
@ronghanghu I've not read read this already but  has it some overlap point with https://github.com/BVLC/caffe/pull/759?

**Comment by ronghanghu** [comment]
@bhack No. This PR is incremental enhancement while #759 involves extending the label format, which may involve many other things (e.g. extending loss layers to support them?) to be taken into consideration.

**Comment by bhack** [comment]
@ronghanghu Ok.. Thank you. So I leave it opened if somebody could be still interested.

File: src/caffe/layers/accuracy_layer.cpp (hunk 2 above)
Comment (rmanor) [inline src/caffe/layers/accuracy_layer.cpp]: Ok, I'll change, by FYI 10k * 2 * sizeof(int) is no more than 160KB :)

File: src/caffe/layers/accuracy_layer.cpp [hunk 4]
```diff
@@ -38,6 +38,10 @@ void AccuracyLayer<Dtype>::Reshape(
       << "with integer values in {0, 1, ..., C-1}.";
   vector<int> top_shape(0);  // Accuracy is a scalar; 0 axes.
   top[0]->Reshape(top_shape);
+  if (top.size() > 1) {
+    int dim = bottom[0]->count() / bottom[0]->num();
+    top[1]->Reshape(dim, 1, 1, 1);
```
Comment (ronghanghu) [inline src/caffe/layers/accuracy_layer.cpp]: A note: in order to better align with ND-Array blob and to be compatible with accuracy axis, please change these two lines into

    vector<int> top_shape2(1, bottom[0]->shape(label_axis_)); // Per-class accuracy is a vector; 1 axes.
    top[1]->Reshape(top_shape2);

And also reshape your buffer blob after here to have `top_shape2`.

**Comment by rmanor** [comment]
Fixed comments, thanks.

**Comment by ronghanghu** [comment]
Please run `make lint` and fix lint errors.

**Comment by rmanor** [comment]
Fixed.

File: src/caffe/layers/accuracy_layer.cpp [hunk 5]
```diff
@@ -82,6 +95,18 @@ void AccuracyLayer<Dtype>::Forward_cpu(const vector<Blob<Dtype>*>& bottom,
...
+    for (int i = 0; i < dim; ++i) {
+      if (has_ignore_label_ && i == ignore_label_) {
+        continue;
+      }
+      top[1]->mutable_cpu_data()[output_label] =
+          nums_buffer_.cpu_data()[i] == 0 ? 0
+          : top[1]->cpu_data()[i] / nums_buffer_.cpu_data()[i];
+      ++output_label;
+    }
+  }
```
Comment (ronghanghu) [inline src/caffe/layers/accuracy_layer.cpp]: I see what you mean here, but it seems confusing to shift label index in `top[1]` due to ignore_label.

Better to keep original index in `top[1]` and simply do

    if (top.size() > 1) {
      int top_count = top[1]->count();
      for (int i = 0; i < top_count; ++i) {
        top[1]->mutable_cpu_data()[i] =
            nums_buffer_.cpu_data()[i] == 0 ? 0
            : top[1]->cpu_data()[i] / nums_buffer_.cpu_data()[i])
      }
    }

Comment (rmanor) [inline src/caffe/layers/accuracy_layer.cpp]: Fixed, thanks.

**Comment by ronghanghu** [comment]
Looks good to me now. Please squash into one single commit :)

**Comment by rmanor** [comment]
Squashed :)

**Comment by ChenSiqin** [comment]
@rmanor Hi, rmanor! Now I use your code to output the accuracy per class. My train_val.prototxt is:

layer {
  name: "accuracy"
  type: "Accuracy"
  bottom: "loss3_animal/classifier"
  bottom: "label"
  top: "accuracy"
  top: "class"
  include {
    phase: TEST
  }
}

and I use the ImageData layer:

layer {
  name: "data"
  type: "ImageData"
  top: "data"
  top: "label"
  include {
    phase: TEST
  }
  transform_param {
    mirror: true
    crop_size: 224
    mean_file: "animal_mean.binaryproto"
  }
  image_data_param {
    source: "/str/users/chensq/experiments/network/caffe/data/39/val_animal_0-37.txt"
    batch_size: 50
    shuffle: true
    new_height: 256
    new_width: 256
  }
}

I am doing 38 animal classes classification problem. In my log file, I find something strange:
I1128 15:04:30.611277  5720 solver.cpp:415]     Test net output #0: accuracy = 0.842105
I1128 15:04:30.611333  5720 solver.cpp:415]     Test net output #1: class = 0.723684
I1128 15:04:30.611343  5720 solver.cpp:415]     Test net output #2: class = 0.559649
I1128 15:04:30.611358  5720 solver.cpp:415]     Test net output #3: class = 0.482456
I1128 15:04:30.611367  5720 solver.cpp:415]     Test net output #4: class = 0.616228
I1128 15:04:30.611374  5720 solver.cpp:415]     Test net output #5: class = 0.679825
I1128 15:04:30.611382  5720 solver.cpp:415]     Test net output #6: class = 0.744737
I1128 15:04:30.611389  5720 solver.cpp:415]     Test net output #7: class = 0.708333
I1128 15:04:30.611397  5720 solver.cpp:415]     Test net output #8: class = 0.629386
I1128 15:04:30.611405  5720 solver.cpp:415]     Test net output #9: class = 0.815789
I1128 15:04:30.611418  5720 solver.cpp:415]     Test net output #10: class = 0.688597
I1128 15:04:30.611428  5720 solver.cpp:415]     Test net output #11: class = 0.330702
I1128 15:04:30.611436  5720 solver.cpp:415]     Test net output #12: class = 0.564912
I1128 15:04:30.611445  5720 solver.cpp:415]     Test net output #13: class = 0.679825
I1128 15:04:30.611454  5720 solver.cpp:415]     Test net output #14: class = 0.379386
I1128 15:04:30.611462  5720 solver.cpp:415]     Test net output #15: class = 0.491228
I1128 15:04:30.611471  5720 solver.cpp:415]     Test net output #16: class = 0.673684
I1128 15:04:30.611480  5720 solver.cpp:415]     Test net output #17: class = 0.701754
I1128 15:04:30.611490  5720 solver.cpp:415]     Test net output #18: class = 0.600877
I1128 15:04:30.611497  5720 solver.cpp:415]     Test net output #19: class = 0.561404
I1128 15:04:30.611506  5720 solver.cpp:415]     Test net output #20: class = 0.701754
I1128 15:04:30.611515  5720 solver.cpp:415]     Test net output #21: class = 0.469298
I1128 15:04:30.611522  5720 solver.cpp:415]     Test net output #22: class = 0.627193
I1128 15:04:30.611531  5720 solver.cpp:415]     Test net output #23: class = 0.665789
I1128 15:04:30.611538  5720 solver.cpp:415]     Test net output #24: class = 0.618421
I1128 15:04:30.611546  5720 solver.cpp:415]     Test net output #25: class = 0.583333
I1128 15:04:30.611554  5720 solver.cpp:415]     Test net output #26: class = 0.574561
I1128 15:04:30.611562  5720 solver.cpp:415]     Test net output #27: class = 0.542982
I1128 15:04:30.611570  5720 solver.cpp:415]     Test net output #28: class = 0.569298
I1128 15:04:30.611578  5720 solver.cpp:415]     Test net output #29: class = 0.58114
I1128 15:04:30.611588  5720 solver.cpp:415]     Test net output #30: class = 0.705263
I1128 15:04:30.611595  5720 solver.cpp:415]     Test net output #31: class = 0.649123
I1128 15:04:30.611603  5720 solver.cpp:415]     Test net output #32: class = 0.434211
I1128 15:04:30.611611  5720 solver.cpp:415]     Test net output #33: class = 0.736842
I1128 15:04:30.611619  5720 solver.cpp:415]     Test net output #34: class = 0.701754
I1128 15:04:30.611627  5720 solver.cpp:415]     Test net output #35: class = 0.682018
I1128 15:04:30.611635  5720 solver.cpp:415]     Test net output #36: class = 0.605263
I1128 15:04:30.611644  5720 solver.cpp:415]     Test net output #37: class = 0.684211
I1128 15:04:30.611651  5720 solver.cpp:415]     Test net output #38: class = 0.710526
I1128 15:04:30.611663  5720 solver.cpp:415]     Test net output #39: loss = 0.811343 (* 1 = 0.811343 loss)

The all per-class accuracy(Test net output #(1-38)) is smaller than the accuracy in the all 38 classes test data(Test net output #0: accuracy). I wonder whether there is something wrong.

**Comment by rmanor** [comment]
@ChenSiqin remember that the final accuracy also depends on how many samples you have from each class, e.g. a class can have zero accuracy, but if it has very little samples then it won't affect much the total accuracy.

**Comment by ChenSiqin** [comment]
@rmanor Sorry to bother you again. My test data has 38 classes, each class has 50 test samples. Look at my log file output, all per-class accuracy is smaller than the final total accuracy, so I am wonder how can the final total accuracy is so high and  it should be at least one class accuracy is higher than the final total accuracy.

**Comment by volgy** [comment]
@chensiqin I guess the discrepancy comes from the behavior of the class-specific accuracy calculation, when there are zero test samples belong to the particular class in the current batch. In that case, the class-accuracy is explicitly zero. Since the reported test accuracy values are averaged across the   batches (test_iter), these (somewhat arbitrary) zero values skew the final results. Very misleading, indeed.

**Comment by laotao** [comment]
How to use this feature in prototxt/python code?

**Comment by TitusPortfolioPage** [comment]
Hey just adding to what laotao asked. How do I implement the same to my prototxt file?

**Comment by pawanhsu** [comment]
Does anyone know how to use this PR? what should i do after i pull this PR???please help!!

**Comment by wlf52520** [comment]
Hi，i wonder to know which caffe version you used for  Output accuracies per class？ I tried to change relevant codes you modified，but failed. could you provide all the caffe codes？ Thank you！
